blark/transform.py

```diff
@@ -1594,7 +1603,9 @@ def from_lark(
         )
 
     def __str__(self) -> str:
-        parameters = ", ".join(str(param) for param in self.parameters)
+        parameters = ""
+        if self.parameters != [None]:
```

Comment:
Hmm, great catch on the bug. I can't believe I haven't run into this yet.

However I think the fix should go in `from_lark` to ensure that the `parameters` list is `[]` and not `[None]`. Just as a double-check - the annotation hints at what was intended - a `List[ParameterAssignment]` and not `List[Optional[ParameterAssignment]]`

Reply:
I think that's what I'd started with, but ran into some failures:

<details>
  <summary>`pytest` Output</summary>

```python-traceback
λ pytest
===================================================================================== test session starts ======================================================================================
platform win32 -- Python 3.9.0, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: C:\Users\joestan\OneDrive - Schweitzer Engineering Laboratories\Python Experiments\1 - SEL AcRTAC Projects\blark
plugins: cov-3.0.0, forked-1.4.0, xdist-2.5.0, xdoctest-0.15.10
collected 323 items

blark\tests\test_parsing.py .........x..............................x                                                                                                                     [ 12%]
blark\tests\test_transformer.py ......................................................................................................................................................... [ 60%]
................x....x...x......................................FF..........................................x.x                                                                           [ 94%]
blark\tests\test_util.py ....x.............                                                                                                                                               [100%]

=========================================================================================== FAILURES =========================================================================================== 
_________________________________________________________ test_statement_roundtrip[if_statement-IF callable() THEN\n    y();\nEND_IF] __________________________________________________________ 

rule_name = 'if_statement', value = 'IF callable() THEN\n    y();\nEND_IF'

    @pytest.mark.parametrize(
        "rule_name, value",
        [
            param("if_statement", tf.multiline_code_block(
                """
                IF 1 THEN
                    iValue := 1;
                    IF 1 THEN
                        iValue := 1;
                    END_IF
                ELSIF 2 THEN
                    iValue := 2;
                ELSIF 3 THEN
                    iValue := 2;
                ELSE
                    iValue := 3;
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF 1 THEN
                    IF 2 THEN
                        IF 3 * x THEN
                            y();
                        ELSE
                        END_IF
                    END_IF
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF callable() THEN
                    y();
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF a() AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN
                    y();
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF 1 AND_THEN 1 THEN
                    y();
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF 0 OR_ELSE 1 THEN
                    y();
                END_IF
                """
            )),
            param("case_statement", tf.multiline_code_block(
                """
                CASE expr OF
                1:
                    abc();
                2, 3, GVL.Constant:
                    def();
                ELSE
                    ghi();
                END_CASE
                """
            )),
            param("case_statement", tf.multiline_code_block(
                """
                CASE a.b.c^.d OF
                1..10:
                    OneToTen := OneToTen + 1;
                EnumValue:
                END_CASE
                """
            )),
            param("case_statement", tf.multiline_code_block(
                """
                CASE expr OF
                BYTE#9..BYTE#10, BYTE#13, BYTE#28:
                    OneToTen := OneToTen + 1;
                EnumValue:
                END_CASE
                """
            )),
            param("case_statement", tf.multiline_code_block(
                """
                CASE expr OF
                TRUE:
                    abc();
                FALSE:
                    def();
                END_CASE
                """
            )),
            param("while_statement", tf.multiline_code_block(
                """
                WHILE expr
                DO
                    iValue := iValue + 1;
                END_WHILE
                """
            )),
            param("repeat_statement", tf.multiline_code_block(
                """
                REPEAT
                    iValue := iValue + 1;
                UNTIL expr
                END_REPEAT
                """
            )),
            param("for_statement", tf.multiline_code_block(
                """
                FOR iIndex := 0 TO 10
                DO
                    iValue := iIndex * 2;
                END_FOR
                """
            )),
            param("for_statement", tf.multiline_code_block(
                """
                FOR iIndex := 0 TO 10 BY 1
                DO
                    iValue := iIndex * 2;
                END_FOR
                """
            )),
            param("for_statement", tf.multiline_code_block(
                """
                FOR iIndex := (iValue - 5) TO iValue * 10 BY iValue MOD 10
                DO
                    arrArray[iIndex] := iIndex * 2;
                END_FOR
                """
            )),
            param("for_statement", tf.multiline_code_block(
                """
                FOR iIndex[1] := 0 TO 10
                DO
                    iValue := iIndex * 2;
                END_FOR
                """
            )),
        ],
    )
    def test_statement_roundtrip(rule_name, value):
>       roundtrip_rule(rule_name, value)

blark\tests\test_transformer.py:934:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

rule_name = 'if_statement', value = 'IF callable() THEN\n    y();\nEND_IF', expected = 'IF callable() THEN\n    y();\nEND_IF'

    def roundtrip_rule(rule_name: str, value: str, expected: Optional[str] = None):
        parser = get_grammar(start=rule_name)
        transformed = parse_source_code(value, parser=parser)
        print("\n\nTransformed:")
        print(repr(transformed))
        print("\n\nOr:")
        print(transformed)
        if expected is None:
            expected = value
>       assert str(transformed) == expected
E       AssertionError: assert 'IF callable(... y();\nEND_IF' == 'IF callable(... y();\nEND_IF'
E         - IF callable() THEN
E         + IF callable(None) THEN
E         ?             ++++
E               y();
E           END_IF

blark\tests\test_transformer.py:35: AssertionError
------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------- 


Transformed:
IfStatement(if_expression=FunctionCall(name=SimpleVariable(name=Token('IDENTIFIER', 'callable'), dereferenced=False), parameters=[None]), statements=StatementList(statements=[MethodStatement(method=SimpleVariable(name=Token('IDENTIFIER', 'y'), dereferenced=False))]), else_ifs=[], else_clause=None)


Or:
IF callable(None) THEN
    y();
END_IF
__________________________________ test_statement_roundtrip[if_statement-IF a() AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN\n    y();\nEND_IF] __________________________________ 

rule_name = 'if_statement', value = "IF a() AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN\n    y();\nEND_IF"

    @pytest.mark.parametrize(
        "rule_name, value",
        [
            param("if_statement", tf.multiline_code_block(
                """
                IF 1 THEN
                    iValue := 1;
                    IF 1 THEN
                        iValue := 1;
                    END_IF
                ELSIF 2 THEN
                    iValue := 2;
                ELSIF 3 THEN
                    iValue := 2;
                ELSE
                    iValue := 3;
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF 1 THEN
                    IF 2 THEN
                        IF 3 * x THEN
                            y();
                        ELSE
                        END_IF
                    END_IF
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF callable() THEN
                    y();
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF a() AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN
                    y();
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF 1 AND_THEN 1 THEN
                    y();
                END_IF
                """
            )),
            param("if_statement", tf.multiline_code_block(
                """
                IF 0 OR_ELSE 1 THEN
                    y();
                END_IF
                """
            )),
            param("case_statement", tf.multiline_code_block(
                """
                CASE expr OF
                1:
                    abc();
                2, 3, GVL.Constant:
                    def();
                ELSE
                    ghi();
                END_CASE
                """
            )),
            param("case_statement", tf.multiline_code_block(
                """
                CASE a.b.c^.d OF
                1..10:
                    OneToTen := OneToTen + 1;
                EnumValue:
                END_CASE
                """
            )),
            param("case_statement", tf.multiline_code_block(
                """
                CASE expr OF
                BYTE#9..BYTE#10, BYTE#13, BYTE#28:
                    OneToTen := OneToTen + 1;
                EnumValue:
                END_CASE
                """
            )),
            param("case_statement", tf.multiline_code_block(
                """
                CASE expr OF
                TRUE:
                    abc();
                FALSE:
                    def();
                END_CASE
                """
            )),
            param("while_statement", tf.multiline_code_block(
                """
                WHILE expr
                DO
                    iValue := iValue + 1;
                END_WHILE
                """
            )),
            param("repeat_statement", tf.multiline_code_block(
                """
                REPEAT
                    iValue := iValue + 1;
                UNTIL expr
                END_REPEAT
                """
            )),
            param("for_statement", tf.multiline_code_block(
                """
                FOR iIndex := 0 TO 10
                DO
                    iValue := iIndex * 2;
                END_FOR
                """
            )),
            param("for_statement", tf.multiline_code_block(
                """
                FOR iIndex := 0 TO 10 BY 1
                DO
                    iValue := iIndex * 2;
                END_FOR
                """
            )),
            param("for_statement", tf.multiline_code_block(
                """
                FOR iIndex := (iValue - 5) TO iValue * 10 BY iValue MOD 10
                DO
                    arrArray[iIndex] := iIndex * 2;
                END_FOR
                """
            )),
            param("for_statement", tf.multiline_code_block(
                """
                FOR iIndex[1] := 0 TO 10
                DO
                    iValue := iIndex * 2;
                END_FOR
                """
            )),
        ],
    )
    def test_statement_roundtrip(rule_name, value):
>       roundtrip_rule(rule_name, value)

blark\tests\test_transformer.py:934:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

rule_name = 'if_statement', value = "IF a() AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN\n    y();\nEND_IF"
expected = "IF a() AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN\n    y();\nEND_IF"

    def roundtrip_rule(rule_name: str, value: str, expected: Optional[str] = None):
        parser = get_grammar(start=rule_name)
        transformed = parse_source_code(value, parser=parser)
        print("\n\nTransformed:")
        print(repr(transformed))
        print("\n\nOr:")
        print(transformed)
        if expected is None:
            expected = value
>       assert str(transformed) == expected
E       AssertionError: assert 'IF a(None) A... y();\nEND_IF' == 'IF a() AND b... y();\nEND_IF'
E         - IF a() AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN
E         + IF a(None) AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN
E         ?      ++++
E               y();
E           END_IF

blark\tests\test_transformer.py:35: AssertionError
------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------- 


Transformed:
IfStatement(if_expression=BinaryOperation(left=BinaryOperation(left=FunctionCall(name=SimpleVariable(name=Token('IDENTIFIER', 'a'), dereferenced=False), parameters=[None]), op=Token('LOGICAL_AND', 'AND'), right=BinaryOperation(left=FunctionCall(name=SimpleVariable(name=Token('IDENTIFIER', 'b'), dereferenced=False), parameters=[InputParameterAssignment(name=SimpleVariable(name=Token('IDENTIFIER', 'in'), dereferenced=False), value=Integer(value=Token('INTEGER', '5'), type_name=None))]), op=Token('EQUALS_OP', '<>'), right=Integer(value=Token('INTEGER', '0'), type_name=None))), op=Token('LOGICAL_AND', 'AND'), right=FunctionCall(name=SimpleVariable(name=Token('IDENTIFIER', 'c'), dereferenced=False), parameters=[InputParameterAssignment(name=SimpleVariable(name=Token('IDENTIFIER', 'in1'), dereferenced=False), value=SimpleVariable(name=Token('IDENTIFIER', 'expr'), dereferenced=False)), InputParameterAssignment(name=SimpleVariable(name=Token('IDENTIFIER', 'in2'), dereferenced=False), value=String(value=Token('SINGLE_BYTE_CHARACTER_STRING', "'test'")))])), statements=StatementList(statements=[MethodStatement(method=SimpleVariable(name=Token('IDENTIFIER', 'y'), dereferenced=False))]), else_ifs=[], else_clause=None)


Or:
IF a(None) AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN
    y();
END_IF
======================================================================================= warnings summary ======================================================================================= 
..\..\..\..\appdata\local\programs\python\python39\lib\site-packages\win32\lib\pywintypes.py:2
  c:\users\joestan\appdata\local\programs\python\python39\lib\site-packages\win32\lib\pywintypes.py:2: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's 
documentation for alternative uses
    import imp, sys, os

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=================================================================================== short test summary info ==================================================================================== 
FAILED blark/tests/test_transformer.py::test_statement_roundtrip[if_statement-IF callable() THEN\n    y();\nEND_IF] - AssertionError: assert 'IF callable(... y();\nEND_IF' == 'IF callable(.....FAILED blark/tests/test_transformer.py::test_statement_roundtrip[if_statement-IF a() AND b(in := 5) <> 0 AND c(in1 := expr, in2 := 'test') THEN\n    y();\nEND_IF] - AssertionError: assert 'I...===================================================================== 2 failed, 313 passed, 8 xfailed, 1 warning in 22.90s =====================================================================
```

</details>

So it still seems a bit... odd... 😕 But let me know what thoughts you have! I'd rather get away from that `None` too!

Reply:
So... perhaps there's something else that I'm negligently overlooking that should *really* be addressed as the root of this?

Reply:
I think this comes in here:
https://github.com/engineerjoe440/blark/blob/ecffb0a5ebe4fa2bb5da4f701dd3b6a8932d5a21/blark/iec.lark#L707
`[ ]` means "give me a `None` if there are no matches to the inner expression in lark (with `maybe_placeholders` configured to be `True`)

So in the case of no parameters, we see `None` come in as the argument in `from_lark` instead of a variable amount of `ParameterAssignment`.

My initial thought is that `from_lark` should then just check for this scenario and convert the argument `parameters = (None, )` to `parameters = []`. It may be possible to clean up the grammar to avoid it, but my brain has yet to get in gear with `iec.lark` this morning...

Reply:
I like your train of thought, on this! Let me poke some bits here and see how it handles 😉 

More soon!

Reply:
Latest commit should contain these changes! 🎉

Reply:
Great - this looks good! Glad I steered you in the right direction. One small suggestion and then this should be good to merge.

I believe the `__str__` changes for `FunctionCall` `parameters` can be reverted now to its more terse form (as in the master branch):

https://github.com/engineerjoe440/blark/blob/49a892e2f7e393928b2bbcda6047b9caa7d21631/blark/transform.py#L1609-L1611